### PR TITLE
Looping Constructs

### PR DESCRIPTION
**Pull request by hinckleyc**
# For/While loops

## Description
This is for the user story about looping constructs. It only checks for and while loops which is what the user wanted.

### Type of Change

- [x] Feature
- [ ] Bug fix
- [ ] Documentation

## Contributors

- @MaddyKapfhammer 
- @BaileyMatrascia 

## Reminder

All GitHub Actions should be in a passing state before any pull request is merged.

### CODE REVIEW

File: hello/test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+"""A program to practice using the LibCST visitor function."""
```
Comment (MaddyKapfhammer) [inline hello/test.py:1]: This file can be removed. We are no longer using the visitor function, depreciating this file.

File: main.py [hunk 2]
```diff
@@ -28,9 +27,13 @@ def if_statements():
...
 @app.command()
-def looping_constructs():
+def looping_constructs(directory_path):
     """Determine number of looping constructs in a Python directory."""
-    print("In progress")
+    cast_dict = generate_trees(directory_path)
+    while_loops_dict = counter.count_whileloops(cast_dict)
+    for_loops_dict = counter.count_forloops(cast_dict)
+
+    print(while_loops_dict, for_loops_dict)
```
Comment (MaddyKapfhammer) [inline main.py:36]: Once the `total_loops` function is completed in `counter.py`, this print statement should be changed to output the total number of looping constructs in a directory.

File: pyproject.toml [hunk 3]
```diff
@@ -18,7 +18,7 @@ pylint = "^2.12.2"
 
 [tool.taskipy.tasks]
 test = "pytest --cov-report term-missing --cov-fail-under=70  --cov-branch --cov=hello tests/"
-lint = "pylint hello/ tests/"
+lint = "pylint tests/"
```
Comment (MaddyKapfhammer) [inline pyproject.toml:21]: Line 20 should read as: `test = "pytest --cov-report term-missing --cov-fail-under=70 --cov-branch --cov castanet tests/"`

Line 21 should read as `lint = "pylint castanet/ tests/"

File: tests/test_counter.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    """Check that functions and docstrings are counted correctly"""
+    directory = "./hello"
+    file_list = generator.find_python_files(directory)
+    string_file_list = generator.read_files(directory, file_list)
+    tree_dict = generator.generate_cast(string_file_list)
+    imports_dictionary1 = counter.count_whileloops(tree_dict)
+    imports_dictionary = counter.count_forloops(tree_dict)
+    assert imports_dictionary1 == {'__init__.py': 0, 'test.py': 0, 'say_hello.py': 1,
+                                   'test-while.py': 0}
+    assert imports_dictionary == ({'__init__.py': 0, 'test.py': 0, 'say_hello.py': 1,
```
Comment (MaddyKapfhammer) [inline tests/test_counter.py:17]: Why are you asserting the same thing twice?

File: tests/test_counter.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+from castanet import generate_trees as generator
+
+
+def test_countloops():
+    """Check that functions and docstrings are counted correctly"""
+    directory = "./hello"
+    file_list = generator.find_python_files(directory)
+    string_file_list = generator.read_files(directory, file_list)
+    tree_dict = generator.generate_cast(string_file_list)
+    imports_dictionary1 = counter.count_whileloops(tree_dict)
```
Comment (MaddyKapfhammer) [inline tests/test_counter.py:13]: Please change names to be more clear about what these variables actually are

File: tests/test_counter.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+"""This module tests the castanet.counter module."""
+
+from castanet import counter
+from castanet import generate_trees as generator
+
```
Comment (MaddyKapfhammer) [inline tests/test_counter.py:5]: You should have test cases determining that the `while_dictionary` and `for_dictionary` contain the proper things. 

There should be additional test cases that test if your total function is working properly.

**Review by MaddyKapfhammer — CHANGES_REQUESTED** [review]

File: castanet/counter.py [hunk 7]
```diff
@@ -2,34 +2,53 @@
...
+        imports_dictionary[file] = len(imports)
+
+    return imports_dictionary1
 
 
-def match_imports(cast_dict):
-    """A function for counting the number of if statements in a Python program."""
+def count_forloops(cast_dict):
+    """A function for coutning the number of while loops in a Python program."""
     imports_dictionary = {}
```
[the remark this answers is not in the document]
Comment (MaddyKapfhammer) [reply castanet/counter.py]: Looks like this was corrected, resolving conversation.

File: pyproject.toml (hunk 3 above)
Comment (MaddyKapfhammer) [reply pyproject.toml:21]: Please review what these lines should look like.

**Comment by hinckleyc** [comment]
I think I fixed all the problems let me know if I missed something.

File: castanet/counter.py [hunk 8]
```diff
@@ -27,6 +26,43 @@ def total_imports(imports_dict):
...
+
+
+def amount_loops(loops_dict):
+    """Find and combine the number of for and while loops in a Python file."""
+    total_loops = 0
+    for file in loops_dict:
+        amount_of_loops = loops_dict[file]
+        total_loops += amount_of_loops
+
+    return total_loops(loops_dict)
```
Comment (MaddyKapfhammer) [inline castanet/counter.py:63]: I believe that this should just read `return total_loops`

File: hello/say_hello.py [hunk 9]
```diff
@@ -8,3 +8,11 @@ def greet(name):
```
Comment (MaddyKapfhammer) [inline hello/say_hello.py:8]: Please revert say_hello.py to how it was before, and remove contents that you added.

File: main.py [hunk 10]
```diff
@@ -25,9 +24,13 @@ def if_statements(directory_path):
 
 
 @app.command()
-def looping_constructs():
+def looping_constructs(directory_path):
     """Determine number of looping constructs in a Python directory."""
-    print("In progress")
+    cast_dict = generate_trees(directory_path)
+    while_loops_dict = counter.count_whileloops(cast_dict)
+    for_loops_dict = counter.count_forloops(cast_dict)
```
Comment (MaddyKapfhammer) [inline main.py:31]: This command should be returning the number of for loops, number of while loops, and those numbers combined. I believe it should look like this:

```
def looping_constructs(directory_path):
    """Determine number of looping constructs in a Python directory."""
    cast_dict = generate_trees(directory_path)
    while_loops_dict = counter.count_whileloops(cast_dict)
    for_loops_dict = counter.count_forloops(cast_dict)
    number_for_loops = counter.amount_loops(for_loops_dict)
    number_while_loops = counter.amount_loops(while_loops_dict)
    total_loops = number_for_loops + number_while_loops
    print("Number for loops: " + number_for_loops)
    print("Number while loops: " + number_while_loops)
    print("Number total looping constructs: " + total_loops)
```

**Review by MaddyKapfhammer — CHANGES_REQUESTED** [review]

**Review by MaddyKapfhammer — APPROVED** [review]
Looks like everything was changed as necessary. Approving this PR for merge to `develop`. Before this PR can be merged, allow @MaddyKapfhammer to resolve merge conflicts for merging `develop` into this branch.

**Review by antlet — APPROVED** [review]

**Pull request merged**